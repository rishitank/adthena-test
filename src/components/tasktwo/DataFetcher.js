import {
  useEffect,
  useState,
  useReducer,
  useRef
} from 'react';

const baseUrl = 'https://jsonplaceholder.typicode.com';

const dataFetchReducer = (state, action) => {
  switch (action.type) {
    case 'FETCH_INIT':
      return {
        ...state,
        isLoading: true,
        error: false
      };
    case 'FETCH_SUCCESS':
      return {
        ...state,
        isLoading: false,
        error: false,
        data: action.payload
      };
    case 'FETCH_ERROR':
      return {
        ...state,
        isLoading: false,
        error: true
      };
    default:
      throw new Error();
  }
};

const useFetch = (initialData, initialUrl = '', first = true) => {
  const firstUpdate = useRef(first);
  const [url, setUrl] = useState(initialUrl);
  const [state, dispatch] = useReducer(dataFetchReducer, {
    isLoading: false,
    error: null,
    data: initialData
  });

  useEffect(() => {
    let cancelFetch = false;

    const fetchData = async () => {
      dispatch({ type: 'FETCH_INIT' });

      try {
        const res = await fetch(url);
        const json = await res.json();

        if (!cancelFetch) {
          dispatch({ type: 'FETCH_SUCCESS', payload: json });
        }
      } catch (e) {
        if (!cancelFetch) {
          dispatch({ type: 'FETCH_ERROR' });
        }
      }
    };

    if (!firstUpdate.current) {
      fetchData();
    }

    firstUpdate.current = false;

    return () => {
      cancelFetch = true;
    };
  }, [url]);

  return [state, setUrl];
};

export { useFetch, baseUrl };
