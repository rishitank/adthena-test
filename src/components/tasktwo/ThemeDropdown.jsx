import React, { useState } from 'react';
import PropTypes from 'prop-types';

const ThemeDropdown = ({ user, themeState: [theme, setTheme] }) => {
  const [items] = useState([
    {
      label: 'light mode',
      value: 'light'
    }, {
      label: 'dark mode',
      value: 'dark'
    }
  ]);

  return (
    <div className="dropdown-holder">
      <h4>select theme:</h4>
      <select
        disabled={user === null || !Object.keys(user).length}
        value={theme}
        className="dropdown"
        onChange={e => setTheme(e.currentTarget.value)}
      >
        {items.map(({ value, label }) => (
          <option
            key={value}
            value={value}
          >
            {label}
          </option>
        ))}
      </select>
    </div>
  );
};

ThemeDropdown.defaultProps = {
  user: null
};

ThemeDropdown.propTypes = {
  user: PropTypes.object,
  themeState: PropTypes.array.isRequired
};

export default ThemeDropdown;
