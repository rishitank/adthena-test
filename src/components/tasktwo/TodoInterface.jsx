import React from 'react';
import PropTypes from 'prop-types';
import { useFetch, baseUrl } from './DataFetcher';

const TodoInterface = ({ userId }) => {
  const [{ data, error, isLoading }] = useFetch([], `${baseUrl}/users/${userId}/todos`, false);

  return (
    <div className="todos">
      <h3><u>Todos</u></h3>
      {error && <div className="message error">Failed to find Todos...</div>}
      {isLoading ? (
        <div className="message">Loading...</div>
      ) : [
        data.length ? (
          <ul>
            {
              data.map(item => (
                <li
                  key={item.id}
                  className={`item ${item.completed ? 'completed' : 'pending'}`}
                >
                  <div>
                    {item.title}
                    {item.completed && (<span>&#10004;</span>)}
                  </div>
                </li>
              ))
            }
          </ul>
        ) : error !== null && (
          <div
            key="error"
            className="message error"
          >
            Todos not found
          </div>
        )
      ]}
    </div>
  );
};

TodoInterface.propTypes = {
  userId: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired
};

export default TodoInterface;
