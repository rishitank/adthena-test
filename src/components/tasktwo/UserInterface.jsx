import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { useFetch, baseUrl } from './DataFetcher';
import ThemeDropdown from './ThemeDropdown';
import TodoInterface from './TodoInterface';

const UserInterface = ({ term }) => {
  const [query, setQuery] = useState(term);
  const themeState = useState('light');
  const [{ data: [user], error, isLoading }, setUrl] = useFetch([]);

  return (
    <div className="user-interface">
      <h2>User Interface</h2>
      <form
        onSubmit={e => {
          setUrl(`${baseUrl}/users?username=${query}`);
          e.preventDefault(); // prevent page reload
        }}
      >
        <input
          type="text"
          placeholder="Enter username"
          value={query}
          onChange={e => setQuery(e.target.value)}
          disabled={isLoading}
          required
        />
        <button type="submit">Search</button>
        <ThemeDropdown
          user={user}
          themeState={themeState}
        />
      </form>
      {error && <div className="message error">Failed to find username...</div>}
      {isLoading ? (
        <div className="message">Loading...</div>
      ) : [
        user ? (
          <div
            key="user-details"
            className={`user-details ${themeState[0]}`}
          >
            <div className="user">
              <h3><u>User Details</u></h3>
              <div>
                <strong>username: </strong>
                {user.username}
              </div>
              <div>
                <strong>email: </strong>
                {user.email}
              </div>
              <div>
                <strong>website: </strong>
                {user.website}
              </div>
            </div>
            <TodoInterface userId={user.id} />
          </div>
        ) : error !== null && (
          <div
            key="error"
            className="message error"
          >
            User not found
          </div>
        )
      ]}
    </div>
  );
};

UserInterface.defaultProps = {
  term: ''
};

UserInterface.propTypes = {
  term: PropTypes.string
};

export default UserInterface;
