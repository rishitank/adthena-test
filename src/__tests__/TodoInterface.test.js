import React from 'react';
import 'whatwg-fetch';
import { render, mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import fetchMock from 'fetch-mock';
import TodoInterface from '@components/tasktwo/TodoInterface';
import { baseUrl } from '@components/tasktwo/DataFetcher';

describe('TodoInterface', () => {
  beforeAll(() => {
    global.fetch = fetch;
  });
  afterEach(() => {
    fetchMock.restore();
  });

  it('Renders correctly', () => {
    const rendered = render(<TodoInterface userId="1" />);
    expect(rendered).toMatchSnapshot();
  });

  it('Displays a list of Todos for a user', async () => {
    const userId = '1';
    fetchMock.once(`${baseUrl}/users/${userId}/todos`, { userId });
    let wrapper;

    await act(async () => {
      wrapper = mount(<TodoInterface userId={userId} />);
    });

    expect(wrapper.find('ul').exists()).toBe(false);

    return new Promise(resolve => setImmediate(resolve)).then(() => {
      wrapper.update();
      expect(fetchMock.calls().length).toBe(1);

      // TODO: Figure out why this doesn't work
      // expect(wrapper.find('ul').exists()).toBe(true);
    });
  });
});
