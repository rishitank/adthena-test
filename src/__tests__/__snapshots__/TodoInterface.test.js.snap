// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoInterface Renders correctly 1`] = `
<div
  class="todos"
>
  <h3>
    <u>
      Todos
    </u>
  </h3>
</div>
`;
